src/log.c

```diff
@@ -133,6 +136,122 @@ void fi_log_init(void)
 	pid = getpid();
 }
 
+static int ofi_log_enabled(int disable, enum fi_log_level level,
+			   enum fi_log_subsys subsys)
+{
+    return ((FI_LOG_TAG(disable, level, subsys) & log_mask) ==
+	    FI_LOG_TAG(disable, level, subsys));
+}
+
+static void ofi_log(const char *buf)
+{
+    fprintf(stderr, "%s", buf);
+}
+
+static int ofi_log_ready(int disable, enum fi_log_level level,
+			 enum fi_log_subsys subsys, uint64_t *showtime);
+
+static struct fi_ops_log log_ops = {
+	.size = sizeof(struct fi_ops),
+	.enabled = ofi_log_enabled,
+	.ready = ofi_log_ready,
+	.log = ofi_log,
+};
+
+static int ofi_close_logging_fid(struct fid *fid);
+static int ofi_bind_logging_fid(struct fid *fid, struct fid *bfid,
+				uint64_t flags);
+
+static struct fi_ops ofi_logging_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = ofi_close_logging_fid,
+	.bind = ofi_bind_logging_fid,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+	.tostr = fi_no_tostr,
+	.ops_set = fi_no_ops_set,
+};
+
+static struct fid_logging log_fid = {
+	.fid = {
+		.fclass = FI_CLASS_LOG,
+		.ops = &ofi_logging_ops,
+	},
+	.ops = &log_ops,
+};
+
+static int ofi_close_logging_fid(struct fid *fid)
+{
+	/* Reset logging ops to default */
+	log_fid.ops->enabled = ofi_log_enabled;
+	log_fid.ops->ready = ofi_log_ready;
+	log_fid.ops->log = ofi_log;
+	log_fid.fid.context = NULL;
+	return 0;
+}
+
+static int ofi_logging_import(struct fid *fid)
+{
+	struct fid_logging *impfid;
+
+	if (fid->fclass != FI_CLASS_LOG)
+		return -FI_EINVAL;
+
+	impfid = container_of(fid, struct fid_logging, fid);
+	if (impfid->ops->size < sizeof(struct fi_ops_log))
+		return -FI_EINVAL;
+
+	log_fid.ops->enabled = impfid->ops->enabled;
```

Comment:
@shefty Do we expect the application to set all the functions in `fi_ops_log`? Or shall we import the functions that are != NULL only?

Reply:
I would allow functions to be NULL and only use what's provided.

Reply:
Done